ERCS/erc-7820.md

```diff
@@ -0,0 +1,411 @@
+
+**RoleGranted(address indexed targetContract, bytes32 indexed role, address indexed account)**: Emitted when a role is granted to an account, this event provides a public log that can be used to track role assignments and changes, ensuring that role grants are transparent and verifiable.
+
+**RoleRevoked(address indexed targetContract, bytes32 indexed role, address indexed account)**: Emitted when a role is revoked from an account, this event similarly ensures that role revocations are publicly logged and traceable, supporting robust access control management.
+
+### Design Decisions
+There are a few design decisions that have to be explicitly specified to ensure the functionality, security, and efficiency of the IAccessControlRegistry:
+
+#### Decentralized Contract Registration
+
+**No Central Owner**: There is no central owner who can register contracts. This design choice promotes decentralization and ensures that individual contracts are responsible for their own registration and management.
+
+#### Efficient Storage and Lookup
+
+**Mapping Utilization**: The use of mappings for storing contract information (mapping(address => ContractInfo) private contracts) and role assignments (mapping(address => mapping(address => mapping(bytes32 => bool))) private _contractRoles) ensures efficient storage and lookup. This is crucial for maintaining performance in a large-scale system with numerous contracts and roles.
+
+#### Role Management Flexibility
+
+**Bulk Operations**: Functions like grantRole and revokeRole allow for the assignment and revocation of roles to multiple accounts for multiple contracts in a single transaction. This bulk operation reduces gas costs and simplifies the process of role management in large systems.
+
+#### Robust Security Measures
+
+**Admin-Only Operations**: Functions that modify the state, such as unRegisterContract, _grantRole, and _revokeRole, are restricted to contract admins. This ensures that only authorized personnel can manage contracts and roles, reducing the risk of unauthorized changes.
+
+**Valid Address Checks**: The validAddress modifier ensures that addresses are non-zero, preventing potential issues with null addresses which could lead to unintended behavior or security vulnerabilities.
+
+**Active Contract Checks**: The onlyActiveContract modifier ensures that actions are only performed on active contracts, preventing operations on inactive or unregistered contracts.
+
+#### Transparent Auditing
+
+**Event Logging**: Emitting events for each significant action (registration, unregistration, role granting, and revocation) provides a transparent log that can be monitored and audited. This helps in detecting and responding to unauthorized or suspicious activities promptly.
+
+## Reference Implementation
+
+```solidity
+// SPDX-License-Identifier: MIT
```

Comment:
Code that's inline in the document itself _must_ be CC0-1.0 licensed. We do allow MIT in the `assets/` directory if you'd like to link to it there.